### PR TITLE
Security/log filter

### PR DESCRIPTION
**Pull request by marcgear**
### Overview

Related GitHub issue: #<tbc>

- [x] Reviewed and filled in QA section
- [x] Reviewed and filled in Code Review section
- [x] Reviewed and filled in Dependencies review section
- [x] Documented development process in PR
- [x] Documented approach and thought process in issue

Added structured JSON logging; generates or picks up a correlation id (x-request-id) for reading aggregated logs.  Supports redaction for anywhere that we need to filter things out of logs that shouldn't be there (like passwords) 

### QA
Verified that log messages appear in the console and that no passwords/secure info is being logged

### Code Review
Some basic logging has been introduced to prove out the concept. 
One thing it doesn't do is pass the correlation id to the platform in shift-node-api.  That needs adding at some point.

### Dependencies Review
New dependency on pino which is a log library designed to be small and quick.  dev dep on pino-pretty which just makes JSON structured logs easier to read in the console on dev.

## ENV variables to be added
new env var:
LOG_LEVEL - trace|debug|info|warn|error|fatal (optional - defaults to warn)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
     "dev": "yarn --cwd packages/reference devserver",
     "cy:ci": "yarn --cwd packages/reference node .circleci/cypress.js",
     "build": "yarn --cwd packages/reference build",
-    "start": "yarn --cwd packages/reference start"
+    "start": "yarn --cwd packages/reference start",
+    "test": "yarn --cwd packages/reference test && yarn --cwd packages/shift-next test && yarn --cwd packages/shift-node-api test && yarn --cwd packages/shift-next-routes test && yarn --cwd packages/shift-react-components test"
```
Comment (cbarratt) [inline package.json]:
```suggestion
    "test": "yarn workspaces run test"
```

File: packages/shift-next-routes/src/express/account-handler.js [hunk 2]
```diff
@@ -36,19 +37,13 @@ module.exports = {
 
       if (response.status === 201) {
         extractCustomerId(req, response.data.data)
+        req.log && req.log.info(`Login by user ${req.session.customerId}`)
```
Comment (cbarratt) [inline packages/shift-next-routes/src/express/account-handler.js]: This data feels more like `req.log.debug`, probably isn't something we care to see for every successful login, just my thoughts though!

File: packages/shift-next-routes/src/express/account-handler.js [hunk 3]
```diff
@@ -249,9 +202,22 @@ function extractCustomerId (req, data) {
   }
 }
 
+function handleErrorResponse(error,req, res) {
```
Comment (cbarratt) [inline packages/shift-next-routes/src/express/account-handler.js]:
```suggestion
function handleErrorResponse(error, req, res) {
```

File: packages/shift-next-routes/src/express/account-handler.js [hunk 4]
```diff
@@ -249,9 +202,22 @@ function extractCustomerId (req, data) {
...
+      return res.status(response.status).send(response.data.errors)
+    default:
+      return res.status(response.status).send(response.data)
+  }
+}
+
 async function assignCartToUser (req, res) {
   if (!req.session.customerId) return
   const cartId = req.signedCookies.cart
+  req.log && req.log.info({ msg: 'assigning cart to user', cartId, customerId: req.session.customerId })
```
Comment (cbarratt) [inline packages/shift-next-routes/src/express/account-handler.js]: totally personal opinion but i believe this should be debug

File: packages/shift-node-api/src/lib/set-cache-headers.js [hunk 5]
```diff
@@ -13,9 +13,9 @@ function setCacheHeaders (response) {
     response.headers['Surrogate-Key'] = surrogateKeys
     response.headers['Surrogate-Control'] = 'max-age=3600,stale-if-error=86400,stale-while-revalidate=86400'
 
-    // print a warning when surrogate keys exceed limit
+    // log a warning when surrogate keys exceed limit
     if (surrogateKeys.length > lengthLimit) {
-      console.log('Warning: The following Surrogate keys have exceeded the fastly length limit:', surrogateKeys)
+      //response.log.warn('The following Surrogate keys have exceeded the fastly length limit:', surrogateKeys)
```
Comment (cbarratt) [inline packages/shift-node-api/src/lib/set-cache-headers.js]: Commented out LOC

File: packages/shift-next-routes/src/express/account-handler.js (hunk 2 above)
Comment (marcgear) [reply packages/shift-next-routes/src/express/account-handler.js]: My argument for this being above debug is that it's a priviledge escalation - I'd normally record all instances of (attempted or successful) priviledge escalation for a service just for auditing purposes.

File: packages/shift-next-routes/src/express/account-handler.js (hunk 4 above)
Comment (marcgear) [reply packages/shift-next-routes/src/express/account-handler.js]: Same reason as above really - audit trail.  but happy to downgrade this one if anyone has strong feelings

**Pull request merged**